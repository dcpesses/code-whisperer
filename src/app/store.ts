import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import counterReducer from '@/features/redux-counter/counterSlice';
import modalCommandListReducer from '@/features/modal-command-list/modalSlice';
// import twitchReducer from '../features/twitch/twitchSlice';
import channelReducer from '@/features/twitch/channel-slice.js';
import userReducer from '@/features/player-queue/user-slice.js';
import queueReducer from '@/features/player-queue/queue-slice';
import todosReducer from '@/features/todos/todosSlice';

const reducer = {
  counter: counterReducer,
  channel: channelReducer,
  modal: modalCommandListReducer,
  queue: queueReducer,
  user: userReducer,
  todos: todosReducer,
};
export const store = configureStore({ reducer });

export function getStoreWithState(preloadedState?: RootState) {
  return configureStore({ reducer, preloadedState });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
