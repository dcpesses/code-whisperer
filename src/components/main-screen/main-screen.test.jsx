/* eslint-env jest */
import { render } from '@testing-library/react';
import {vi} from 'vitest';
import React from 'react';
import { Provider } from 'react-redux';
import { getStoreWithState } from '@/app/store';
import { LOCALSETTINGS_KEY } from '@/features/twitch/settings-slice';
import MessageHandler from '@/features/twitch-messages/message-handler';
import MainScreen, {MainScreen as MainScreenComponent, noop} from './index';

vi.mock('@/api/twitch');
vi.mock('../../../package.json', () => {
  return {
    ...vi.importActual('../../../package.json'),
    version: '0.0.0'
  };
});

const getMockTwitchApi = (overrides={}) => Object.assign({
  isChatConnected: true,
  _chatClient: {},
  onMessage: vi.fn(),
  sendMessage: vi.fn().mockResolvedValue([]),
  updateLastMessageTime: vi.fn(),
}, overrides);

const getMessageHandlerConfig = (overrides={}) => Object.assign({
  access_token: 'mockAccessToken',
  addGameRequest: vi.fn(),
  allowGameRequests: true,
  // changeNextGameIdx: this.changeNextGameIdx,
  channel: 'mockBroadcaster',
  clearQueueHandler: vi.fn(),
  closeQueueHandler: vi.fn(),
  joinQueueHandler: vi.fn(),
  listQueueHandler: vi.fn(),
  logUserMessages: false,
  messages: {},
  moderators: [],
  onDelete: vi.fn(),
  onMessageCallback: vi.fn(),
  onSettingsUpdate: vi.fn(),
  openQueueHandler: vi.fn(),
  playerExitHandler: vi.fn(),
  // previousGames: this.state.history.slice(0, this.state.nextGameIdx),
  removeSelectedGameFromHistory: vi.fn(),
  setNextGame: vi.fn(),
  settings: {},
  startGame: vi.fn(),
  toggleAllowGameRequests: vi.fn(),
  twitchApi: vi.fn().mockReturnValue(getMockTwitchApi()),
  upcomingGames: [],
  debug: false,
  init: false,
}, overrides);

const storeState = {
  channel: {
    user: {
      id: '1',
      login: 'twitchstreamer',
      display_name: 'TwitchStreamer',
      type: '',
      broadcaster_type: '',
      description: 'description',
      profile_image_url: 'https://static-cdn.jtvnw.net/jtv_user_pictures/profile_image-300x300.png',
      offline_image_url: '',
      view_count: 0,
      created_at: '2019-11-18T00:47:34Z'
    },
    moderators: [{
    }],
    vips:[{
    }],
  },
  user: {
    info: {
      id: '0',
      login: 'twitchuser',
      display_name: 'TwitchUser',
      type: '',
      broadcaster_type: '',
      description: 'description',
      profile_image_url: 'https://static-cdn.jtvnw.net/jtv_user_pictures/profile_image-300x300.png',
      offline_image_url: '',
      view_count: 0,
      created_at: '2019-11-18T00:47:34Z'
    },
    chatters: {
      twitchuser: {
        id: '0',
        login: 'twitchuser',
        display_name: 'TwitchUser',
        type: '',
        broadcaster_type: '',
        description: 'description',
        profile_image_url: 'https://static-cdn.jtvnw.net/jtv_user_pictures/profile_image-300x300.png',
        offline_image_url: '',
        view_count: 0,
        created_at: '2019-11-18T00:47:34Z'
      }
    },
    moderatedChannels: [{
      broadcaster_id: '1',
      broadcaster_login: 'TwitchStreamer',
      broadcaster_name: 'twitchstreamer',
    }],
    whisperStatus: {
      twitchuser: {
        login: 'twitchuser',
        response: {
          msg: 'Code sent to @TwitchUser',
          status: 204
        }
      },
    }
  }
};

const mockLSGetItem = (key) => {
  switch (key) {
  case LOCALSETTINGS_KEY:
    return JSON.stringify({customDelimiter: '|'});
  case '__version':
    return '1.0.0';
  default:
    return null;
  }
};

describe('noop', () => {
  test('should execute without error', () => {
    expect(noop()).toBeUndefined();
  });
});


describe('MainScreen', () => {

  describe('constructor', () => {
    let updateAppSettings;

    beforeEach(() => {
      updateAppSettings = vi.fn();
    });

    test('should load and initialize settings from localStorage', () => {
      vi.spyOn(window.localStorage.__proto__, 'getItem').mockImplementation(mockLSGetItem);
      const mainScreen = new MainScreenComponent({updateAppSettings});
      expect(updateAppSettings).toBeCalledWith({
        'customDelimiter': '|',
        'enableRoomCode': true,
      });
      expect(mainScreen).toBeDefined();
    });
    test('should handle any errors that occur when initializing settings and use the default settings', () => {
      vi.spyOn(window.localStorage.__proto__, 'getItem').mockReturnValue('{]');
      const mainScreen = new MainScreenComponent({updateAppSettings});
      expect(updateAppSettings).not.toHaveBeenCalled();
      expect(mainScreen).toBeDefined();
    });
  });

  describe('componentDidMount', () => {
    let mainScreen;

    beforeEach(() => {
      vi.spyOn(window.localStorage.__proto__, 'getItem').mockImplementation(mockLSGetItem);
      mainScreen = new MainScreenComponent({
        setFakeUserStates: vi.fn(),
        setFakeChannelStates: vi.fn(),
        setFakeSettingsStates: vi.fn(),
        showOnboarding: vi.fn(),
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'initMessageHandler');
    });

    test('should initialize messageHandler if TwitchApi is available', () => {
      mainScreen.props.twitchApi = {
        isChatConnected: true
      };
      mainScreen.componentDidMount();
      expect(mainScreen.initMessageHandler).toHaveBeenCalled();
    });

    test('should not initialize messageHandler if twitchApi is not available', () => {
      mainScreen.props.twitchApi = null;
      mainScreen.initMessageHandler = vi.fn();
      mainScreen.componentDidMount();
      expect(mainScreen.initMessageHandler).not.toHaveBeenCalled();
    });

    test('should load fake state data', () => {
      vi.spyOn(window.location, 'hash', 'get').mockReturnValue('?fakestate=true');

      mainScreen.componentDidMount();
      expect(mainScreen.props.setFakeUserStates).toHaveBeenCalled();
      expect(mainScreen.props.setFakeChannelStates).toHaveBeenCalled();
      expect(mainScreen.props.setFakeSettingsStates).toHaveBeenCalled();
    });

    test('should show onboarding prompt modal for first time user', () => {
      vi.spyOn(window.localStorage.__proto__, 'getItem').mockImplementation(
        (key) => {
          switch (key) {
          case LOCALSETTINGS_KEY:
            return JSON.stringify({customDelimiter: true});
          default:
            return null;
          }
        }
      );
      mainScreen = new MainScreenComponent({
        setFakeUserStates: vi.fn(),
        setFakeChannelStates: vi.fn(),
        setFakeSettingsStates: vi.fn(),
        showOnboarding: vi.fn(),
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'initMessageHandler');
      mainScreen.componentDidMount();
      expect(mainScreen.state.showOnboardingPromptModal).toBeTruthy();
      // expect(mainScreen.props.showOnboarding).toHaveBeenCalled();
    });

    test('should not show onboarding for past users', () => {
      mainScreen.componentDidMount();
      expect(mainScreen.state.showOnboardingPromptModal).toBeFalsy();
      // expect(mainScreen.props.showOnboarding).not.toHaveBeenCalled();
    });
  });

  describe('componentDidUpdate', () => {
    let mainScreen;

    beforeEach(() => {
      mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'initMessageHandler').mockReturnValue({});
      vi.spyOn(mainScreen, 'updateMessageHandler').mockReturnValue(null);
    });

    test('should initialize messageHandler when chat client connection changes', () => {
      mainScreen.messageHandler = new MessageHandler(getMessageHandlerConfig());
      mainScreen.props = {
        twitchApi: {
          isChatConnected: true
        }
      };
      const prevProps = {
        twitchApi: {
          isChatConnected: false
        }
      };
      mainScreen.componentDidUpdate(prevProps, {});
      expect(mainScreen.initMessageHandler).toHaveBeenCalled();
      expect(mainScreen.updateMessageHandler).not.toHaveBeenCalled();
    });

    test('should initialize messageHandler when chat client is connected but messageHandler not yet set', () => {
      mainScreen.messageHandler = null;
      mainScreen.props = {
        twitchApi: {
          isChatConnected: true
        }
      };
      const prevProps = {
        twitchApi: {
          isChatConnected: true
        }
      };
      mainScreen.componentDidUpdate(prevProps, {});
      expect(mainScreen.initMessageHandler).toHaveBeenCalled();
      expect(mainScreen.updateMessageHandler).not.toHaveBeenCalled();
    });

    test('should update messageHandler when chat client is connected and does not change AND remove any custom commands when unset', () => {
      mainScreen.messageHandler = {
        updateChatCommandTerm: vi.fn()
      };
      mainScreen.props = {
        settings: {
          customJoinCommand: null,
          customLeaveCommand: null,
          customQueueCommand: null,
        },
        twitchApi: {
          isChatConnected: true
        }
      };

      const prevProps = {
        settings: {
          customJoinCommand: '!joinme',
          customLeaveCommand: '!leaveme',
          customQueueCommand: '!queueme',
        },
        twitchApi: {
          isChatConnected: true
        }
      };

      mainScreen.componentDidUpdate(prevProps);
      expect(mainScreen.updateMessageHandler).toHaveBeenCalled();
      expect(mainScreen.messageHandler.updateChatCommandTerm).toHaveBeenCalledTimes(3);
      expect(mainScreen.initMessageHandler).not.toHaveBeenCalled();
    });
    test('should not update messageHandler if messageHandler does not exist and chat client not yet connected', () => {
      mainScreen.props = {
        twitchApi: {
          isChatConnected: false
        }
      };
      const prevProps = {
        twitchApi: {
          isChatConnected: false
        }
      };
      mainScreen.componentDidUpdate(prevProps);
      expect(mainScreen.initMessageHandler).not.toHaveBeenCalled();
      expect(mainScreen.updateMessageHandler).not.toHaveBeenCalled();
    });

    test('should not update messageHandler if twitchApi is not available', () => {
      mainScreen.props = {
        twitchApi: null
      };
      const prevProps = {};
      mainScreen.componentDidUpdate(prevProps);
      expect(mainScreen.initMessageHandler).not.toHaveBeenCalled();
      expect(mainScreen.updateMessageHandler).not.toHaveBeenCalled();
    });
  });

  describe('initMessageHandler', () => {
    let mainScreen;

    beforeEach(() => {
      mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
    });

    test('should return null if twitchApi is not available', () => {
      mainScreen.props = { twitchApi: null };
      const messageHandler = mainScreen.initMessageHandler();
      expect(messageHandler).toBe(null);
    });

    test('should initialize messageHandler correctly if twitchApi is available', () => {
      mainScreen.props = {
        settings: {
          customJoinCommand: '!joinme',
          customLeaveCommand: '!leaveme',
          customQueueCommand: '!queueme',
        },
        twitchApi: getMockTwitchApi()
      };
      const messageHandler = mainScreen.initMessageHandler();
      expect(messageHandler).toBeTruthy();
    });
  });

  describe('updateMessageHandler', () => {
    let mainScreen;

    beforeEach(() => {
      mainScreen = new MainScreenComponent({
        twitchApi: getMockTwitchApi(),
        updateAppSettings: vi.fn(),
      });
    });
    test('should update messageHandler if both messageHandler and TwitchApi available', () => {
      vi.spyOn(mainScreen, 'initMessageHandler');
      mainScreen.props.access_token = 'mockUpdatedAccessToken';
      mainScreen.props.channel = 'TwitchChannel';
      mainScreen.state.logUserMessages = true;
      mainScreen.state.messages = { 'gameObj-longName': {} };
      mainScreen.props.moderators = { player1: {id: '11'} };
      mainScreen.props.settings = {
        customJoinCommand: '!joinme',
        customLeaveCommand: '!leaveme',
        customQueueCommand: '!queueme',
      };
      expect(mainScreen.updateMessageHandler()).toBeNull();
      expect(mainScreen.initMessageHandler).toHaveBeenCalledTimes(0);
    });
    test('should call initMessageHandler and return if chat client is connected but only accessible via props.twitchApi', () => {
      vi.spyOn(mainScreen, 'initMessageHandler');
      mainScreen.props.twitchApi = getMockTwitchApi();
      mainScreen.twitchApi = { isChatConnected: false };
      expect(mainScreen.updateMessageHandler({ twitchApi: getMockTwitchApi() })).toBeUndefined();
      expect(mainScreen.initMessageHandler).toHaveBeenCalledTimes(1);
    });
    test('should call initMessageHandler and return if chat client is not connected', () => {
      vi.spyOn(mainScreen, 'initMessageHandler');
      mainScreen.messageHandler = null;
      mainScreen.props.twitchApi = {};
      mainScreen.twitchApi = {};
      expect(mainScreen.updateMessageHandler()).toBeUndefined();
      expect(mainScreen.initMessageHandler).toHaveBeenCalledTimes(0);
    });

  });

  describe('onMessageHandlerInit', () => {
    let mainScreen;
    let messageHandler;

    beforeEach(() => {
      messageHandler = new MessageHandler(getMessageHandlerConfig());
      mainScreen = new MainScreenComponent({
        messageHandler,
        twitchApi: getMockTwitchApi(),
        updateAppSettings: vi.fn(),
      });
      mainScreen.messageHandler = messageHandler;
    });

    test('should update chat command terms if settings are updated', () => {
      mainScreen.props = {
        settings: {
          customJoinCommand: '!customJoin',
          customLeaveCommand: '!customLeave',
          customQueueCommand: '!customQueue',
          enableRestrictedListQueue: true
        }
      };
      vi.spyOn(messageHandler, 'updateChatCommandTerm'); // TODO: migrate all to updateChatCommand
      vi.spyOn(messageHandler, 'updateChatCommand');

      mainScreen.onMessageHandlerInit();

      expect(messageHandler.updateChatCommandTerm).toHaveBeenCalledTimes(3);
      expect(messageHandler.updateChatCommandTerm).toHaveBeenCalledWith('joinQueue', '!customJoin');
      expect(messageHandler.updateChatCommandTerm).toHaveBeenCalledWith('leaveQueue', '!customLeave');
      expect(messageHandler.updateChatCommandTerm).toHaveBeenCalledWith('listQueue', '!customQueue');
      expect(messageHandler.updateChatCommand).toHaveBeenCalledTimes(4);
      expect(messageHandler.updateChatCommand).toHaveBeenCalledWith('joinQueue', 'commands', '!customJoin');
      expect(messageHandler.updateChatCommand).toHaveBeenCalledWith('leaveQueue', 'commands', '!customLeave');
      expect(messageHandler.updateChatCommand).toHaveBeenCalledWith('listQueue', 'commands', '!customQueue');
      expect(messageHandler.updateChatCommand).toHaveBeenCalledWith('listQueue', 'mod', true);

    });

    test('should not update chat command terms if settings are not updated', () => {
      mainScreen.props = { settings: null };
      vi.spyOn(mainScreen.messageHandler, 'updateChatCommandTerm');

      mainScreen.onMessageHandlerInit();

      expect(messageHandler.updateChatCommandTerm).not.toHaveBeenCalled();
    });

    test('should return if messageHandler is not set', () => {
      mainScreen.messageHandler = null;

      expect(mainScreen.onMessageHandlerInit()).toBeUndefined();
    });
  });

  describe('getOptionsDebugMenu', () => {
    let mainScreen;
    let optionsDebugMenu;

    beforeEach(() => {
      mainScreen = new MainScreenComponent({
        setFakeQueueStates: vi.fn(),
        setFakeUserStates: vi.fn(),
        setFakeChannelStates: vi.fn(),
        updateAppSettings: vi.fn(),
      });
      optionsDebugMenu = mainScreen.getOptionsDebugMenu();
    });
    test('should be defined', () => {
      expect(optionsDebugMenu).toBeDefined();
      expect(optionsDebugMenu.every(item => !!item.label && !!item.onClick)).toBeTruthy();
    });
    test('should load mock player requests', () => {
      let menuItem = optionsDebugMenu.find(item => item.label === 'Load Mock Player Requests');
      expect(menuItem).toBeDefined();
      expect(menuItem.onClick()).not.toBeDefined();
      expect(mainScreen.props.setFakeQueueStates).toHaveBeenCalled();
      expect(mainScreen.props.setFakeUserStates).toHaveBeenCalled();
      expect(mainScreen.props.setFakeChannelStates).toHaveBeenCalled();
    });
    test('should log the debug environment', () => {
      let menuItem = optionsDebugMenu.find(item => item.label === 'Log Debug Environment');
      expect(menuItem).toBeDefined();
      expect(menuItem.onClick()).not.toBeDefined();
    });
    test('should toggle the user message logging', () => {
      vi.spyOn(mainScreen, 'toggleUserMessageLogging');
      let menuItem = optionsDebugMenu.find(item => item.label === 'Toggle User Message Logging');
      expect(menuItem).toBeDefined();
      expect(menuItem.onClick()).not.toBeDefined();
      expect(mainScreen.toggleUserMessageLogging).toHaveBeenCalled();
    });
  });

  describe('toggleUserMessageLogging', () => {
    test('should toggle the state of logUserMessages', () => {
      const mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'setState');
      mainScreen.state.logUserMessages = false;
      mainScreen.toggleUserMessageLogging();
      expect(mainScreen.setState.mock.calls[0][0](mainScreen.state)).toEqual({ logUserMessages: true });
    });
  });

  describe('onMessage', () => {
    let mainScreen;
    let props;

    beforeEach(() => {
      props = {
        userLookup: {},
        setUserLookup: vi.fn(),
        setChatterInfo: vi.fn(),
        twitchApi: getMockTwitchApi({
          updateLastMessageTime: vi.fn(),
          requestUserInfo: vi.fn().mockResolvedValue({ data: [{ login: 'testUser' }] })
        }),
        updateAppSettings: vi.fn(),
      };
      mainScreen = new MainScreenComponent(props);
      mainScreen.twitchApi = props.twitchApi;
    });

    test('should update user lookup and chatter info when user metadata is provided', async() => {
      const message = 'Hello!';
      const user = 'testUser';
      const metadata = { 'user-id': '123' };
      await mainScreen.onMessage(message, user, metadata);
      expect(mainScreen.props.setUserLookup).toHaveBeenCalledWith(metadata);
      expect(mainScreen.props.setChatterInfo).toHaveBeenCalledWith({ login: 'testUser' });
    });

    test('should not update user lookup and chatter info when user metadata is not provided', async() => {
      const message = 'Hello!';
      const user = 'testUser';
      await mainScreen.onMessage(message, user);
      expect(mainScreen.props.setUserLookup).not.toHaveBeenCalled();
      expect(mainScreen.props.setChatterInfo).not.toHaveBeenCalled();
    });
  });

  describe('onSettingsUpdate', () => {
    let mainScreen;

    beforeEach(() => {
      vi.spyOn(window.localStorage.__proto__, 'setItem');
      mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
      mainScreen.updateMessageHandler = vi.fn();
      mainScreen.props.updateAppSettings = vi.fn();
    });

    test('should update settings in state, save to localStorage, and call updateMessageHandler', () => {
      const nextSettings = { enableRoomCode: false };

      mainScreen.onSettingsUpdate(nextSettings);

      expect(localStorage.setItem).toHaveBeenCalledWith('__settings', JSON.stringify(nextSettings));
      expect(mainScreen.props.updateAppSettings).toHaveBeenCalled();
      expect(mainScreen.updateMessageHandler).toHaveBeenCalled();
    });
  });

  describe('onShowOnboarding', () => {
    test('should hide modal and show onboarding walkthrough', () => {
      const mainScreen = new MainScreenComponent({
        showOnboarding: vi.fn(),
        updateAppSettings: vi.fn(),
      });
      mainScreen.state.showOnboardingPromptModal = true;
      vi.spyOn(mainScreen, 'setState');
      mainScreen.onShowOnboarding();
      expect(mainScreen.props.showOnboarding).toHaveBeenCalled();
      expect(mainScreen.setState.mock.calls[0][0](mainScreen.state)).toEqual({ showOnboardingPromptModal: false });
    });
  });

  describe('toggleChangelogModal', () => {
    test('should toggle the state of showChangelogModal', () => {
      const mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'setState');
      mainScreen.state.showChangelogModal = false;
      mainScreen.toggleChangelogModal();
      expect(mainScreen.setState.mock.calls[0][0](mainScreen.state)).toEqual({ showChangelogModal: true });
    });
  });

  describe('toggleOptionsMenu', () => {
    test('should toggle the state of toggleOptionsMenu', () => {
      const mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'setState');
      mainScreen.state.showOptionsMenu = false;
      mainScreen.toggleOptionsMenu();
      expect(mainScreen.setState.mock.calls[0][0](mainScreen.state)).toEqual({ showOptionsMenu: true });
    });
  });

  describe('toggleOnboardingPromptModal', () => {
    test('should toggle the state of showOnboardingPromptModal', () => {
      const mainScreen = new MainScreenComponent({
        updateAppSettings: vi.fn(),
      });
      vi.spyOn(mainScreen, 'setState');
      mainScreen.state.showOnboardingPromptModal = false;
      mainScreen.toggleOnboardingPromptModal();
      expect(mainScreen.setState.mock.calls[0][0](mainScreen.state)).toEqual({ showOnboardingPromptModal: true });
    });
  });

  describe('sendWhisper', () => {
    let mainScreen;
    let props;

    beforeEach(() => {
      props = {
        setWhisperStatus: vi.fn(),
        twitchApi: {
          sendWhisper: vi.fn().mockResolvedValue()
        },
        updateAppSettings: vi.fn(),
      };
      mainScreen = new MainScreenComponent(props);
    });

    test('should call sendWhisper method and update whisper status when TwitchApi is available', async() => {
      const player = { username: 'testUser' };
      const msg = '!join';
      await mainScreen.sendWhisper(player, msg);
      expect(mainScreen.props.twitchApi.sendWhisper).toHaveBeenCalledWith(player, msg);
      expect(mainScreen.props.setWhisperStatus).toHaveBeenCalledWith({ login: player.username, response: undefined });
    });

    test('should handle cases when twitchApi.sendWhisper is not available', async() => {
      mainScreen.twitchApi = null;
      const player = { username: 'testUser' };
      const msg = '!join';
      const resp = await mainScreen.sendWhisper(player, msg);
      expect(resp).toBeUndefined();
    });
  });


  describe('render', () => {
    let store;
    let twitchApi;
    beforeEach(() => {
      store = getStoreWithState(storeState);
      twitchApi = getMockTwitchApi({
        isChatConnected: true,
        updateLastMessageTime: vi.fn(),
        requestUserInfo: vi.fn().mockResolvedValue({ data: [{ login: 'testUser' }] })
      });
    });
    test('Should render without error', () => {
      const {container} = render(
        <Provider store={store}>
          <MainScreen twitchApi={twitchApi} />
        </Provider>
      );
      expect(container).toMatchSnapshot();
    });
  });
});
