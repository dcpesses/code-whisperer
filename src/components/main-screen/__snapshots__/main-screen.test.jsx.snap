// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainScreen > Should render without error 1`] = `
<div>
  <div
    class="main-screen"
  >
    <nav
      class="bg-body-tertiary mb-3 py-0 raleway-font navbar navbar-light"
      data-bs-theme="dark"
    >
      <div
        class="container-fluid"
      >
        <span
          class="fw-semibold navbar-brand"
        >
          <img
            alt="TwitchUser"
            class="rounded-circle navbar-pfp-img proxy-profile-img"
            src="https://static-cdn.jtvnw.net/jtv_user_pictures/profile_image-300x300.png"
          />
          <img
            alt="TwitchStreamer"
            class="rounded-circle navbar-pfp-img"
            src="https://static-cdn.jtvnw.net/jtv_user_pictures/profile_image-300x300.png"
          />
           
          TwitchStreamer
        </span>
        <button
          aria-controls="navbar-options-menu"
          aria-label="Toggle navigation"
          class="border-0 rounded-0 navbar-toggler collapsed"
          data-margin-right=""
          style=""
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          aria-labelledby="navbar-options-menu-label"
          class="raleway-font offcanvas offcanvas-end"
          id="navbar-options-menu"
        >
          <div
            class="offcanvas-header"
          >
            <div
              class="fw-bold fs-4 offcanvas-title h5"
              id="navbar-options-menu-label"
            >
              Options
            </div>
            <button
              aria-label="Close"
              class="btn-close"
              type="button"
            />
          </div>
          <div
            class="fw-medium offcanvas-body"
          >
            <div
              class="justify-content-end flex-grow-1 pe-3 fs-5 navbar-nav"
            >
              <a
                class="nav-link"
                href="#"
                role="button"
                tabindex="0"
              >
                Logout
              </a>
              <hr
                class="border-bottom my-2"
              />
              <a
                class="settings-menu nav-link"
                href="#"
                role="button"
                tabindex="0"
              >
                Settings
              </a>
              <div
                class="accordion-dark accordion accordion-flush collapse"
                id="settings-menu"
              >
                <div
                  class="accordion-body"
                >
                  <button
                    class="btn settings-menu btn btn-link"
                    id="enable-room-code"
                    title="Allows host to set a room code that can be whispered to players."
                    type="button"
                  >
                    <input
                      readonly=""
                      role="switch"
                      type="checkbox"
                    />
                     
                    <span>
                      Enable Room Code
                    </span>
                  </button>
                  <button
                    class="btn settings-menu btn btn-link"
                    title="Replaces the !join command with a custom term to use to join the Interested queue."
                    type="button"
                  >
                    <span
                      for="custom-join-command"
                    >
                      Use Custom Join Command: 
                    </span>
                    <input
                      class="form-control"
                      id="custom-join-command"
                      name="custom-join-command"
                      placeholder="e.g. !join"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                  </button>
                  <button
                    class="btn settings-menu btn btn-link"
                    title="Messages users in chat when they've successfully joined the Interested queue."
                    type="button"
                  >
                    <input
                      readonly=""
                      role="switch"
                      type="checkbox"
                    />
                     
                    <span>
                      Show Join Confirmation Message
                    </span>
                  </button>
                  <button
                    class="btn settings-menu btn btn-link"
                    title="Replaces the !leave command with a custom term to use to leave all of the queues."
                    type="button"
                  >
                    <span
                      for="custom-leave-command"
                    >
                      Use Custom Leave Command: 
                    </span>
                    <input
                      class="form-control"
                      id="custom-leave-command"
                      name="custom-leave-command"
                      placeholder="e.g. !leave"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                  </button>
                  <button
                    class="btn settings-menu btn btn-link"
                    title="Messages users in chat when they've successfully left the queues."
                    type="button"
                  >
                    <input
                      readonly=""
                      role="switch"
                      type="checkbox"
                    />
                     
                    <span>
                      Show Leave Confirmation Message
                    </span>
                  </button>
                  <button
                    class="btn settings-menu btn btn-link"
                    title="Uses a custom character or emote to separate requests listed in the chat."
                    type="button"
                  >
                    <span
                      for="custom-delimiter"
                    >
                      Use Custom Delimiter: 
                    </span>
                    <input
                      class="form-control"
                      id="custom-delimiter"
                      name="custom-delimiter"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                  </button>
                  <hr
                    class="border-bottom my-2"
                  />
                  <h5>
                    Beta Options
                  </h5>
                  <div
                    class="smaller"
                  >
                    These options have not been fully tested and may not work as intended.
                  </div>
                  <button
                    class="btn settings-menu btn btn-link"
                    id="enable-moderated-channels-option"
                    title="Allows user to use this app on another channel that grants them moderation access."
                    type="button"
                  >
                    <input
                      readonly=""
                      role="switch"
                      type="checkbox"
                    />
                     
                    <span>
                      Enable Moderated Channels Menu
                    </span>
                  </button>
                </div>
              </div>
              <a
                class="nav-link"
                href="#"
                role="button"
                tabindex="0"
              >
                View Chat Commands
              </a>
              <a
                class="nav-link"
                href="#"
                role="button"
                tabindex="0"
              >
                What's New
              </a>
              <div
                class="position-absolute bottom-0 start-0 end-0 pb-3 text-center"
                id="options-debug-menu-items"
              >
                <a
                  class="nav-link"
                  href="#"
                  role="button"
                  tabindex="0"
                  title="Wanna support the development of Code Whisperer? Donations are never expected but are always appreciated!"
                >
                  <img
                    alt="Support Me on Ko-fi"
                    src="https://ko-fi.com/img/githubbutton_sm.svg"
                  />
                </a>
                <div
                  class="dropup-center dropup"
                  id="dropdown-debug-menu-items"
                  variant="link"
                >
                  <button
                    aria-expanded="false"
                    class="text-decoration-none dropdown-toggle btn btn-link btn-sm"
                    id="dropdown-debug-menu-items-toggle"
                    type="button"
                  >
                    version 0.0.0
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="container mx-auto"
      id="content"
    >
      <div
        class="queues d-flex flex-column flex-md-row my-2 flex-wrap"
      >
        <div
          class="queue my-1 px-md-1 col-12"
        >
          <div
            class="d-flex align-items-center p-3 my-3 text-white bg-purple rounded shadow-sm libre-franklin-font col-lg-8 col-xl-7 col-xxl-6 mx-auto"
          >
            <div
              class="input-group"
            >
              <input
                aria-autocomplete="none"
                aria-describedby="btn-send-to-queue"
                aria-label="Enter Room Code"
                autocomplete="false"
                class="form-control libre-franklin-font"
                name="room-code"
                placeholder="ENTER ROOM CODE"
                role="presentation"
                title="Paste Room Code Here"
                type="text"
                value=""
              />
              <button
                class="btn btn-primary libre-franklin-font text-uppercase"
                disabled=""
                id="btn-send-to-queue"
                title="Send Code to All Players"
                type="button"
              >
                Send to Queue
              </button>
            </div>
          </div>
        </div>
        <div
          class="queue my-1 px-md-1 col-12"
        >
          <div
            class="bg-body rounded shadow-sm p-2 d-flex justify-content-between"
          >
            <div
              class="toggle-streamer-seat"
            >
              <label
                class="toggle-label form-check-label"
                for="reserve-seat-for-streamer"
              >
                Reserve seat for streamer?
              </label>
              <div
                class="form-check form-switch"
              >
                <input
                  class="form-check-input"
                  id="reserve-seat-for-streamer"
                  role="switch"
                  type="checkbox"
                />
              </div>
            </div>
            <div
              class="fs-6 lh-sm align-self-center"
            >
              <div
                class="player-count"
              >
                0
                 of 
                8
                 seats claimed
              </div>
            </div>
          </div>
        </div>
        <div
          class="queue my-1 px-md-1 col-12 col-md-6 order-2 order-md-1"
        >
          <div
            class="bg-body rounded shadow-sm p-2"
          >
            <h6
              class="pb-2 m-2 mb-0 libre-franklin-font text-dark-emphasis text-uppercase clearfix d-flex align-items-bottom"
            >
              <span
                class="queue-header me-auto align-self-center"
              >
                <i
                  class="bi-people text-purple-1 fs-5"
                />
                 Interested
              </span>
              <button
                class="btn btn-sm"
              >
                Randomize
              </button>
            </h6>
            <div
              class="d-flex flex-column text-body interested-queue rand-0"
            />
          </div>
        </div>
        <div
          class="queue my-1 px-md-1 col-12 col-md-6 order-1 order-md-2"
        >
          <div
            class="bg-body rounded shadow-sm p-2"
          >
            <h6
              class="pb-2 m-2 mb-0 libre-franklin-font text-dark-emphasis text-uppercase clearfix d-flex align-items-bottom"
            >
              <span
                class="queue-header me-auto align-self-center"
              >
                <i
                  class="bi-people-fill text-purple-1 fs-5"
                />
                 Playing
              </span>
              <button
                class="btn btn-sm strt-game disabled"
              >
                Clear Seats
              </button>
            </h6>
            <div
              class="d-flex flex-column text-body playing-queue"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
